tpm2/credential_activation.go

```diff
@@ -0,0 +1,134 @@
+package tpm2
+
+import (
+	"crypto"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/google/go-tpm/tpmutil"
+)
+
+// GenerateCredentialActivation creates a wrapped secret for use in credential activation.
+// This function implements the minimum necessary to compute activations for RSA EKs
+// compliant with TCG 2.0 EK Credential Profile specification, revision 14.
+// This should be all windows8+ cerified hardware implementing a TPM 2.0.
+// Specifically, the constraints are:
+// - aik must be a digest rather than a handle, computed using SHA256.
+// - pub must be a 2048 bit RSA key, representing a compliant EK. That means using
+//   SHA256 for digests, and a 128 bit AES symmetric cipher.
+// - symBlockSize should represent the block size of the symmetric cipher. The only
+//   valid value at this time is 16.
+// - secret must not be longer than the longest digest size implemented by the TPM.
+//   A 32 byte secret is a safe, recommended default.
+//
+// This function implements Credential Protection as defined in section 24 of the TPM
+// specification revision 2 part 1, with the additional restrictions listed above.
+// See: https://trustedcomputinggroup.org/resource/tpm-library-specification/
+func GenerateCredentialActivation(aik *Name, pub crypto.PublicKey, symBlockSize int, secret []byte) ([]byte, error) {
```

Comment:
This blob is `TPM2B_ID_OBJECT || TPM2B_ENCRYPTED_SECRET`. I can represent it better & use a struct to serialize it, but its a completely opaque/useless structure to implement a decoder for as it only makes sense to the target TPM.

Reply:
Yes, please make 2 separate structs with exported encoders. It's ok to leave out decoder for now, until it's needed.

Reply:
Will do.

Reply:
`TPM2B_ENCRYPTED_SECRET` is just a `[]byte` with a length prefix, do we really want to make a struct for that? Users can just call `tpmutil.Pack()` on their `[]byte`.

Reply:
Right. You can make a struct for just `TPMS_ID_OBJECT`.
I'd leave it to the user to combine the two. Otherwise the returned blob is a partial `TPM2_ActiateCredential` request

Reply:
Oky doky!

Reply:
I had something a bit different in mind:
- write _separate_ constructor funcs for IDObject and encrypted secret
- remove GenerateCredentialActivation, caller would just use the above two separate constructors
- let them generate the seed

Usage would look something like:
```
// generate seed
encryptedSecret, err := tpm2.NewEncryptedSecret(aik, pub, rnd, seed)
// handle err
idObject, err := tpm2.NewIDObject(aik, secret, seed, encryptedSecret)
// handle err
credentialActivation, err := tpmutil.Pack(encryptedSecret, idObject)
// handle err
```

It's more work for the caller, but keeps the tpm2 code nicely separated.
The above logic can be wrapped in a helper func somewhere in caller's code or a new library.

FYI @tmroeder brought up the idea of separating pure TPM2 serialization and commands from helper funcs into distinct packages. I'll toy around with that when there's free time, but avoiding tight coupling like this PR would help in that process.

Reply:
Whats the benefit of making the caller do all this? I would understand if the `TPM2B_ID_OBJECT` or `TPM2B_ENCRYPTED_SECRET` could be used for other things than just creating an activation, but there is only one use case, which is credential activation.

Further, both `TPM2B_ENCRYPTED_SECRET` & `TPM2B_ID_OBJECT` are opaque structs with no useful information for contexts outside of the TPM (this is intended for anonymity), so its not like they can be used on their own.

Splitting them up makes sense if either the encrypted secret or the ID object had a use apart from ActivateCredential, but in the absence of that it seems like unnecessary complexity?

Reply:
It deviates from the rest of the library by returning what is essentially a partial request.
We should only return encoded individual TPM types.

The code in your func is useful, but IMO it's the wrong place for it. At this level of abstraction we shouldn't implement helpers for TPM-based flows, if possible.
Same as `os/exec` shouldn't have helpers for wrapping commands in a shell, even though that's a very common use case.

Reply:
Where should this function be then? tpmutil?

Reply:
The intent of this function is to generate a challenge on a server - where you dont have a TPM device but need to work very closely with TPM2 structures. A client could always just use `MakeCredential`, and would never need to call `KDFa()` as the TPM handles all that mess.

I'm happy to put it in a new package, but the coupling will be much, much higher thanks to hash alg details + hashConstructor map not being exported.

Reply:
In caller's codebase ideally. Or a new library outside of go-tpm.

Reply:
We already have `ActivateCredential`, `GenerateCredentialActivation` would be confusing to have alongside